### PR TITLE
#263 refactor QR code scanner

### LINKED ISSUE #263: Cleanup qr code parsing error handling
While working on #261 it again became clear that the current error handling when parsing identification and verification qr codes is ugly to read and hard to maintain.

This concerns these two files:
https://github.com/ehrenamtskarte/ehrenamtskarte/blob/main/frontend/lib/identification/identification_qr_content_parser.dart
and
https://github.com/ehrenamtskarte/ehrenamtskarte/blob/main/frontend/lib/verification/scanner/verification_qr_content_parser.dart

### PR DESCRIPTION
**Pull request by elKei24**
Closes #263

### CODE REVIEW

File: frontend/lib/identification/identification_qr_content_parser.dart [hunk 1]
```diff
@@ -1,114 +1,88 @@
...
+    final base64Decoder = Base64Decoder();
+
+    CardActivateModel cardActivateModel;
     try {
-      final base64Decoder = Base64Decoder();
+      var rawProtobufData = base64Decoder.convert(rawBase64Content);
+      cardActivateModel = CardActivateModel.fromBuffer(rawProtobufData);
+    } on Exception catch (_) {
+      throw QRCodeInvalidFormatException();
+    }
```
Comment (RoAnBu) [inline frontend/lib/identification/identification_qr_content_parser.dart:41]: Just wondering, what kind of stuff could happen during protobuf parsing? Could it be worth it printing out an error message to the console?

File: frontend/lib/identification/identification_qr_content_parser.dart [hunk 2]
```diff
@@ -1,114 +1,88 @@
...
-                "der nicht für die Ehrenamtskarte App generiert wurde.");
-      }
+    final fullName = cardActivateModel.fullName;
+    if (fullName == null) {
+      throw QrCodeFieldMissingException("fullName");
+    }
+    final hashSecret = cardActivateModel.hashSecret;
+    if (hashSecret == null) {
+      throw QrCodeFieldMissingException("randomBytes");
+    }
```
Comment (RoAnBu) [inline frontend/lib/identification/identification_qr_content_parser.dart:50]: Do you know if there is a difference between `cardActivateModel.hasFullName()` and the null check we are currently doing? 
Should we also make simple length checks like size > 0?

File: frontend/lib/identification/identification_qr_scanner_page.dart [hunk 3]
```diff
@@ -0,0 +1,51 @@
+import 'package:flutter/cupertino.dart';
```
Comment (RoAnBu) [inline frontend/lib/identification/identification_qr_scanner_page.dart:1]:
```suggestion
```

File: frontend/lib/qr_code_scanner/qr_code_processor.dart [hunk 4]
```diff
@@ -0,0 +1,13 @@
+class QrCodeParseException implements Exception {
+  final String reason;
+  QrCodeParseException(this.reason);
+  String toString() => "${runtimeType.toString()}: $reason";
+}
+
+class QrCodeFieldMissingException extends QrCodeParseException {
+  String missingFieldName;
```
Comment (RoAnBu) [inline frontend/lib/qr_code_scanner/qr_code_processor.dart:8]:
```suggestion
  final String missingFieldName;
```

File: frontend/lib/qr_code_scanner/qr_code_scanner_page.dart [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+      if (onCodeScanned != null) {
+        await onCodeScanned(code);
+      }
+    } on QrCodeParseException catch (e, stackTrace) {
+      debugPrintStack(stackTrace: stackTrace, label: e?.toString());
+    }
+
+    Navigator.of(context).maybePop();
+  }
+}
```
Comment (RoAnBu) [inline frontend/lib/qr_code_scanner/qr_code_scanner_page.dart:38]: Is it intended that we leave the qr code scanner on error?

File: frontend/lib/verification/view/negative_verification_result.dart [hunk 6]
```diff
@@ -10,37 +10,12 @@ class NegativeVerificationResult extends StatelessWidget {
...
-                          .apply(fontSizeFactor: 1.5, fontWeightDelta: 2)),
-                  verificationError.errorTextSpan,
-                  TextSpan(
-                      text: "\nFehlercode: ${verificationError.errorCode}"),
-                ])),
-      ]),
-    );
+        Text(verificationError.errorText),
+        Text("Fehlercode: ${verificationError.errorCode}"),
+      ]));
```
Comment (RoAnBu) [inline frontend/lib/verification/view/negative_verification_result.dart:19]: Font gets really small with just Text(...), hardly readable on ios

File: frontend/lib/verification/view/positive_verification_result.dart [hunk 7]
```diff
@@ -18,31 +18,14 @@ class PositiveVerificationResult extends StatelessWidget {
...
-          TextSpan(text: "Landkreis: ${cardDetails.regionId}"),
-        ])));
+    return VerificationResultCard(
+        title: "Die Ehrenamtskarte ist gültig",
+        isPositive: true,
+        child: Column(
+          children: [
+            Text("Name: ${cardDetails.fullName}"),
+            Text("Ablaufdatum: $expirationDateString"),
+            Text("Landkreis: ${cardDetails.regionId}"),
```
Comment (RoAnBu) [inline frontend/lib/verification/view/positive_verification_result.dart:28]: Again font size

File: frontend/lib/verification/view/verification_view.dart [hunk 8]
```diff
@@ -63,16 +64,43 @@ class _VerificationViewState extends State<VerificationView> {
...
+
+  Future<void> _handleQrCode(String rawQRContent) async {
+    // TODO still no beautiful function
+    final provider =
+    Provider.of<VerificationCardDetailsModel>(context, listen: false);
+    void handleError(String message, String errorCode, Exception e) {
+      print("Verification failed: ${e.toString()}");
+      provider.setVerificationFailure(VerificationError(message,
+          errorCode));
+    }
```
Comment (RoAnBu) [inline frontend/lib/verification/view/verification_view.dart:85]: wait, a function in a function? 🤔

File: frontend/lib/verification/view/verification_result_card.dart [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+  Widget build(BuildContext context) {
+    var icon = isPositive ? Icons.verified_user : Icons.error;
+    var color = isPositive ? Colors.green : Colors.red;
+    return ContentCard(
+        borderSide: isPositive
+            ? const BorderSide(width: 1.0, color: Colors.black12)
+            : const BorderSide(color: Colors.red, width: 4.0),
+        child: Column(
+        children: [
+          RichText( // TODO feels strange to use RichText. Why not Row?
```
Comment (RoAnBu) [inline frontend/lib/verification/view/verification_result_card.dart:27]: tbh, not 100% sure but that's what richtext is doing, I just took that from the docs 😆

**Review by RoAnBu — APPROVED** [review]
Very nice rework!

File: frontend/lib/identification/identification_qr_content_parser.dart (hunk 1 above)
Comment (elKei24) [reply frontend/lib/identification/identification_qr_content_parser.dart:41]: Good point, I will add that 👍🏽 
- [ ] log error

File: frontend/lib/identification/identification_qr_content_parser.dart (hunk 2 above)
Comment (elKei24) [reply frontend/lib/identification/identification_qr_content_parser.dart:50]: Also good points. I kept that logic just as it was, but makes sense to think about those thinks when already touching it.
- [ ] check difference `hasFullName()` and `== null`
- [ ] check for empty strings

File: frontend/lib/qr_code_scanner/qr_code_scanner_page.dart (hunk 5 above)
Comment (elKei24) [reply frontend/lib/qr_code_scanner/qr_code_scanner_page.dart:38]: Well, it is, but I would also be happy with leaving it open. What do you prefer?

File: frontend/lib/verification/view/negative_verification_result.dart (hunk 6 above)
Comment (elKei24) [reply frontend/lib/verification/view/negative_verification_result.dart:19]: Oh, okay, I already wondered why the font is set to be so big 😆

File: frontend/lib/verification/view/verification_view.dart (hunk 8 above)
Comment (elKei24) [reply frontend/lib/verification/view/verification_view.dart:85]: I found this a bit shorter than putting print/setFailure commands in every catch clause

File: frontend/lib/verification/view/verification_result_card.dart (hunk 9 above)
Comment (elKei24) [reply frontend/lib/verification/view/verification_result_card.dart:27]: Ah, okay 😄:+1:

File: frontend/lib/qr_code_scanner/qr_code_scanner_page.dart (hunk 5 above)
Comment (RoAnBu) [reply frontend/lib/qr_code_scanner/qr_code_scanner_page.dart:38]: For testing it was always useful that it remained open. However, idk if that is useful in practice. Probably not? I mean it's easy to change if we need to

File: frontend/lib/identification/identification_qr_content_parser.dart (hunk 1 above)
Comment (Schedulaar) [reply frontend/lib/identification/identification_qr_content_parser.dart:41]: Minor: Is it a best practice to include "inner exceptions" or is that bad style? :see_no_evil:

File: frontend/lib/identification/identification_qr_content_parser.dart [hunk 10]
```diff
@@ -1,114 +1,88 @@
...
-                "werden. Vermutlich handelt es sich um einen QR Code, "
-                "der nicht für die Ehrenamtskarte App generiert wurde.");
-      }
+    final fullName = cardActivateModel.fullName;
+    if (fullName == null) {
+      throw QrCodeFieldMissingException("fullName");
+    }
+    final hashSecret = cardActivateModel.hashSecret;
+    if (hashSecret == null) {
+      throw QrCodeFieldMissingException("randomBytes");
```
Comment (Schedulaar) [inline frontend/lib/identification/identification_qr_content_parser.dart:49]:
```suggestion
      throw QrCodeFieldMissingException("hashSecret");
```

**Review by Schedulaar — APPROVED** [review]
There will be some merge conflicts with #277

File: frontend/lib/qr_code_scanner/qr_code_scanner_page.dart (hunk 5 above)
Comment (elKei24) [reply frontend/lib/qr_code_scanner/qr_code_scanner_page.dart:38]: Okay. I will just keep it as is for now then

File: frontend/lib/identification/identification_qr_content_parser.dart (hunk 1 above)
Comment (elKei24) [reply frontend/lib/identification/identification_qr_content_parser.dart]: In my opinion, it is good style to do so when the inner exception carries relevant information 👍🏽

File: frontend/lib/identification/identification_qr_content_parser.dart (hunk 2 above)
Comment (elKei24) [reply frontend/lib/identification/identification_qr_content_parser.dart]: > Do you know if there is a difference between cardActivateModel.hasFullName() and the null check we are currently doing?

I looked it up. The `has` method for basic types also only checks for `null`. On lists, it also checks that the list is not empty.

I will use the `has` methods. Improves readability and they must be there for a reason 😄

File: frontend/lib/verification/view/negative_verification_result.dart (hunk 6 above)
Comment (elKei24) [reply frontend/lib/verification/view/negative_verification_result.dart]: I did a UI rework in #279 so I would keep it as is in this PR, if you are okay with that.

**Pull request merged**